processor/deltatocumulativeprocessor/internal/streams/expiry.go

```diff
@@ -0,0 +1,60 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package streams // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/streams"
+
+import (
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/identity"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/staleness"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/exp/metrics/streams"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/clock"
+)
+
+func ExpireAfter[T any](items streams.Map[T], ttl time.Duration) Expiry[T] {
+	exp := Expiry[T]{
+		Staleness: *staleness.NewStaleness[T](ttl, items),
+		sig:       make(chan struct{}),
+	}
+	return exp
+}
+
+var _ streams.Map[any] = (*Expiry[any])(nil)
+
+type Expiry[T any] struct {
+	staleness.Staleness[T]
+	sig chan struct{}
+}
+
+func (e Expiry[T]) ExpireOldEntries() <-chan struct{} {
+	e.Staleness.ExpireOldEntries()
+
+	n := e.Staleness.Len()
+	sig := make(chan struct{})
+
+	go func() {
+		switch {
+		case n == 0:
+			<-e.sig
+		case n > 0:
+			expires := e.Staleness.Next().Add(e.Max)
```

Comment:
This has the potential of not being protected by the lock. Since you release the lock on return of this function, but this is a go-routine.

IMO, this whole struct could go away, and the Processor could keep `Staleness` structs at the top level instead of maps. While "elegant", I'm not sure the elegance of sleeping "exactly" to the next expiry outweighs the potential bugs like this race, and the extra logic you had to add to Store() to make sure you can preemptively wake things up. 

The handling of `n` and calculation of `expires` needs to happen outside the goroutine, so it's protected by the lock

Reply:
My vote would be to replace this logic with a simple Ticker() every say 5 minutes. (We *could* make that a config value if want)

Reply:
that's a really good catch and it says quite something about complexity that no one noticed it before.

The issue I see with a timer is that streams possibly live up to `2*max-1` seconds if the timer hits in a bad moment

Reply:
I believe it would be `max+tickerInterval` seconds. IMO in most real-world scenarios `max` will be >> the ticker value. IE, `max` is 1 hour, and ticker is 5 minutes. So a series lasting 1h 4.9999min isn't the end of the world.

There is precedent for this kind of expiry schema in other metrics tools, like [mtail](https://github.com/google/mtail/blob/main/docs/Deploying.md#setting-garbage-collection-intervals)

We *could* allow the tickerInterval to be configurable. And add config validation that `max` > `tickerInterval`. With a big doc / warning that says that `tickerInterval should be much smaller than max to ensure a bad lineup of the tick doesn't allow series to last too long`